maven_iuvs/echelle.py

```diff
@@ -841,3 +880,1530 @@ def find_files_with_geometry(file_index):
+    clean_data : boolean 
+                 whether to perform the data cleaning routines
+
+    Returns:
+    ----------
+    spectrum : array
+               Spectrum in total DN summed over the spatial dimension
+    
+    """
+    # Collect pixel range which we need to find the slit start and end 
+    si1, si2 = get_ech_slit_indices(light_fits)
+
+    if coadded:
+        spectrum = np.sum(data[si1:si2, :], axis=0)
+    else:
+        # Sum up the spectra over the range in which Ly alpha is visible on the slit (not outside it)
+        # This spectrum is thus in DN
+        spectrum = np.sum(data[integration, si1:si2, :], axis=0)
+        
+    if average:
+        spectrum = spectrum / (si2 - si1)
+
+    return spectrum
+
+
+def add_in_quadrature(uncertainties, light_fits, integration=0): 
+    """
+    Similar to get_spectrum, but adds up the uncertainties in what is hopefully 
+    the correct manner. 
+
+    Parameters:
+    ----------
+    data : array
+           3D numpy array of image detector data, dark subtracted and cleaned. 
+    light_fits : astropy.io.fits instance
+                 File with light observation
+    integration : int
+                  Integration frame to use for the specrum. Used if coadded=False.
+    clean_data : boolean 
+                 whether to perform the data cleaning routines
+
+    Returns:
+    ----------
+    spectrum : array
+               Spectrum in total DN summed over the spatial dimension
+    
+    """
+    # Collect pixel range which we need to find the slit start and end 
+    si1, si2 = get_ech_slit_indices(light_fits)
+
+    total_uncert = np.sqrt( np.sum( (uncertainties[integration, si1:si2, :])**2, axis=0) )
+
+    return total_uncert
+
+
+def ran_DN_uncertainty(light_fits, dark_subtracted_and_cleaned_data):
+    """
+    Figure out the random uncertainty in DN.
+
+    # Let's start by just wholesale adapting the uncertainty calculation from Matteo in the IDL pipeline and see what it looks like.
+    # In progress
+    """
+    volt = mcp_dn_to_volt(light_fits['Engineering'].data['MCP_GAIN'][0])   # I think it's wrong to use this. IDL is acquiring the voltage. mcp_volt_to_gain(, channel="FUV")
+    n_bins = light_fits['primary'].header['spe_size'] * light_fits['primary'].header['spa_size'] # in a square bin of spatial x spectral. works out to 22 for recent dat
+    sigma_background = 4313 * math.sqrt(light_fits["Primary"].header["INT_TIME"]/60) * math.sqrt(n_bins/480.)/(2**((850-volt)/50.)) # WHERE the heck did I get this??
+    fit_function = 40 / (2**((700-volt)/50))
+    
+    # This is the correct shape, not sure if it's reasonable values though:
+    ran_DN = np.sqrt(dark_subtracted_and_cleaned_data * fit_function + sigma_background**2) # TODO: check if it's okay to do this on the cleaned data. Probably not
+    ran_DN[np.where(np.isnan(ran_DN))] = 0 # TODO: this is not acceptable lol
+
+    return ran_DN
+
+
+def get_wavelengths(light_fits):
```

Comment:
move to instrument submodule?

Reply:
Hmm, I'll move it there for now, but it does need to be updated to work with not-echelle. I remember that the wavelengths shift a lot on each spatial row in the standard resolution mode.

Reply:
Fixed by [328034e](https://github.com/lasp/maven_iuvs/pull/66/commits/328034e06050588e666afc743d054fb1c0b91e9b), which introduced a circular import problem, which was resolved with [e414214](https://github.com/lasp/maven_iuvs/pull/66/commits/e4142144b46e1c7036a33f0afaba9e97b59c4d85) and [2305251](https://github.com/lasp/maven_iuvs/pull/66/commits/2305251ba468bc8d571e8f937062101b12d7cdfb).